models/Asset.php

```diff
@@ -1525,83 +1525,95 @@ public function removeMetadata(string $name, ?string $language = null)
         return $this;
     }
 
-    /**
-     * @param string|null $name
-     * @param string|null $language
-     * @param bool $strictMatch
-     * @param bool $raw
-     *
-     * @return array|string|null
-     */
-    public function getMetadata($name = null, $language = null, $strictMatch = false, $raw = false)
+    public function getMetadata(?string $name = null, ?string $language = null, bool $strictMatchLanguage = false, bool $raw = false): array|string|null
     {
         $preEvent = new AssetEvent($this);
         $preEvent->setArgument('metadata', $this->metadata);
         $this->dispatchEvent($preEvent, AssetEvents::PRE_GET_METADATA);
         $this->metadata = $preEvent->getArgument('metadata');
 
-        $convert = function ($metaData) {
-            $loader = Pimcore::getContainer()->get('pimcore.implementation_loader.asset.metadata.data');
-            $transformedData = $metaData['data'];
+        if ($name) {
+            return $this->getMetadataByName($name, $language, $strictMatchLanguage, $raw);
+        }
 
-            try {
-                /** @var Data $instance */
-                $instance = $loader->build($metaData['type']);
-                $transformedData = $instance->transformGetterData($metaData['data'], $metaData);
-            } catch (UnsupportedException $e) {
-            }
+        $metaData = $this->getObjectVar('metadata');
+        $result = [];
+        $metaDataWithLanguage = [];
+        if (is_array($metaData)) {
+            foreach ($metaData as $md) {
+                $md = (array)$md;
 
-            return $transformedData;
-        };
+                if ((empty($md['language']) && !$strictMatchLanguage) || ($language == $md['language']) || !$language) {
+                    if (!$raw) {
+                        $md['data'] = $this->transformMetadata($md);
+                    }
+                    $result[] = $md;
+                }
 
-        if ($name) {
-            if ($language === null) {
-                $language = Pimcore::getContainer()->get(LocaleServiceInterface::class)->findLocale();
+                if(!empty($md['language'])) {
+                    $metaDataWithLanguage[$md['language']][$md['name']] = $md;
+                }
             }
+        }
 
-            $data = null;
-            foreach ($this->metadata as $md) {
-                if ($md['name'] == $name) {
-                    if ($language == $md['language']) {
-                        if ($raw) {
-                            return $md;
-                        }
+        if ($language && !$strictMatchLanguage) {
+            foreach($result as $key => &$item) {
+                $lang = $item['language'];
```

Comment:
```suggestion

```
this line can be deleted.

Please also add a upgrade note about the improvement.